subprojects/language-groovy/src/main/java/org/gradle/api/tasks/compile/GroovyCompile.java

```diff
@@ -281,10 +281,10 @@ private static void validateIncrementalCompilationOptions(List<File> sourceRoots
     private GroovyJavaJointCompileSpec createSpec() {
         DefaultGroovyJavaJointCompileSpec spec = new DefaultGroovyJavaJointCompileSpecFactory(compileOptions).create();
 
-        List<File> sourceRoots = CompilationSourceDirs.inferSourceRoots((FileTreeInternal) getSource());
+        List<File> sourceRoots = CompilationSourceDirs.inferSourceRoots((FileTreeInternal) getStableSources().getAsFileTree());
 
         spec.setSourcesRoots(sourceRoots);
-        spec.setSourceFiles(getSource());
+        spec.setSourceFiles(getStableSources().getAsFileTree());
```

Comment:
I think we should call `getStableSources()` once here too.

Reply:
Done.